libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -994,36 +1007,30 @@ void _IotMqtt_DecrementConnectionReferences( _mqttConnection_t * pMqttConnection
 IotMqttError_t IotMqtt_Init( void )
 {
     IotMqttError_t status = IOT_MQTT_SUCCESS;
+    uint32_t initializationStatus = Atomic_CompareAndSwap_u32( &_initCalled,
+                                                               MQTT_LIBRARY_INITIALIZED,
+                                                               MQTT_LIBRARY_UNINITIALIZED );
 
-    if( _initCalled == 0 )
+    if( initializationStatus != 0 )
     {
         /* Call any additional serializer initialization function if serializer
          * overrides are enabled. */
         #if IOT_MQTT_ENABLE_SERIALIZER_OVERRIDES == 1
             #ifdef _IotMqtt_InitSerializeAdditional
                 if( _IotMqtt_InitSerializeAdditional() == false )
                 {
+                    /* Log initialization status. */
+                    IotLogError( "Failed to initialize MQTT library serializer. " );
+
                     status = IOT_MQTT_INIT_FAILED;
                 }
```

Comment:
This is true, however if we try to make this not be the case we'll raise other issues.  For example, if `_IotMqtt_InitSerializeAdditional()` fails, do we allow a call to `IotMqtt_Cleanup()`?  What will `_IotMqtt_CleanupSerializeAdditional()` do in that case?  Will it fail?  If it does, we will have no way of knowing that it did because it [doesn't return a status.](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/doc/lib/mqtt.txt#L190)

I think the safest course of action for the time being is to simply say that calling any MQTT library API after a failed call to `IotMqtt_Init()` will result in undefined behavior.

Reply:
I'm fine with that. However, nit: if the serializer init fails, then the log will look like:
```
[ERROR] Failed to initialize MQTT library serializer. 
[INFO] MQTT library successfully initialized.
```
and the function will return `IOT_MQTT_INIT_FAILED`. This could be a bit confusing, so I think either the info log could be in an if block, or setting the status could be in a set-and-goto.

Reply:
Good point, fixed.